doc/type_erasure.md

```diff
@@ -0,0 +1,590 @@
+# Type Erasure in libunifex
+
+Type erasure is an important tool for managing compile-times by insulating different pieces of code from
+each other.  and also for implementing polymorphism
+
+The libunifex library provides several generic type-erasing wrappers built on top of `tag_invoke`-based customisation point objects (CPOs).
+These generic wrappers can be parameterised on a variadic list of CPOs that define the operations that concrete types must
+support and that the type-erased wrapper itself will expose.
+
+## Categories of type-erasing wrappers
+
+There are several different axes on which different type-erasing wrappers can make different decisions:
+* **Ownership** - Does the wrapper own the wrapped concrete object?
+  * **Unique Ownership** - The wrapper exclusively owns the lifetime of the concrete object it's wrapping.
+  * **Shared Ownership** - The wrapper has shared ownership of the concrete object it's wrapping.
+  * **No Ownership** - The lifetime of the wrapped object is managed outside of the wrapper.
+* **Object Storage** - How does the wrapper store the wrapped object?
+  * **Heap Allocated** - Wrapper stores the concrete object on the heap (possibly allocated using an allocator).
+  * **Stored In-Place** - Wrapper stores the concrete object in-place, inline within the storage of the wrapper object.
+  * **Hybrid Storage** - Wrapper stores small objects inline in the wrapper and larger objects on the heap. Commonly referred to as small-object optimisation.
+  * **External Storage** - Wrapper stores a reference/pointer to the object which is stored in memory not managed by the wrapper.
+* **VTable Storage** - How and where is the VTable for the type-erased object stored?
+  * **Indirect VTable** - Wrapper stores a pointer to the statically allocated vtable. Pointer indirection to call the method.
+  * **In-Place VTable** - Wrapper stores vtable entries in-line inside the wrapper.
+    Avoids pointer-indirection at cost of larger wrapper types. Also allows casting other type-erased types to wrapper types with narrower types.
+  * **Hybrid VTable** - Wrapper stores vtable inline for small vtables and stores it out-of-line for large vtables.
+* **Copyability** - Whether the wrapper is copyable, move-only or non-movable.
+  * **Copyable** - Wrapper has a copy-constructor and move-constructor (may be either `noexcept(true)` or `noexcept(false)`)
+  * **Move-Only** - Wrapper has move-constructor but no copy-constructor (my be either `noexcept(true)` or `noexcept(false)`)
```

Comment:
Where does this leave types that implement a throwing copy constructor (e.g. because it allocates some resource) but that do not have a move-constructor implemented and so the move-constructor falls back to the throwing copy-constructor?

> When dynamic allocation is allowed, types with throwing move constructors should live on the heap.

This is a discussion that I was having with @ispeters before.
When move-constructors are required to be nothrow is it preferable to not support types that don't have noexcept move constructors or to paper over their lack of nothrow move ctor by heap-allocating them?
I guess it seems like you are on the side of "make it work even if it's at the cost of a heap allocation".

Reply:
> Where does this leave types that implement a throwing copy constructor (e.g. because it allocates some resource) but that do not have a move-constructor implemented and so the move-constructor falls back to the throwing copy-constructor?

You mean buggy types. I don't care to support buggy types. Fix your types.

Reply:
Ok, I've updated it so that it heap allocates the value if either the concrete type is not nothrow-move-constructible or if the inline storage is too small or doesn't have the sufficient alignment.